flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierHandler.java

```diff
@@ -19,34 +19,31 @@
 package org.apache.flink.streaming.runtime.io;
 
 import org.apache.flink.annotation.Internal;
+import org.apache.flink.runtime.io.AsyncDataInput;
 import org.apache.flink.runtime.io.network.partition.consumer.BufferOrEvent;
 import org.apache.flink.runtime.jobgraph.tasks.AbstractInvokable;
 
 import java.io.IOException;
+import java.util.Optional;
+
+import static org.apache.flink.util.Preconditions.checkState;
 
 /**
  * The CheckpointBarrierHandler reacts to checkpoint barrier arriving from the input channels.
  * Different implementations may either simply track barriers, or block certain inputs on
  * barriers.
  */
 @Internal
-public interface CheckpointBarrierHandler {
-
+public interface CheckpointBarrierHandler extends AsyncDataInput<Optional<BufferOrEvent>> {
 	/**
-	 * Returns the next {@link BufferOrEvent} that the operator may consume.
-	 * This call blocks until the next BufferOrEvent is available, or until the stream
-	 * has been determined to be finished.
-	 *
-	 * @return The next BufferOrEvent, or {@code null}, if the stream is finished.
-	 *
-	 * @throws IOException Thrown if the network or local disk I/O fails.
-	 *
-	 * @throws InterruptedException Thrown if the thread is interrupted while blocking during
-	 *                              waiting for the next BufferOrEvent to become available.
-	 * @throws Exception Thrown in case that a checkpoint fails that is started as the result of receiving
-	 *                   the last checkpoint barrier
+	 * Blocking version of {@link #pollNext()}.
 	 */
-	BufferOrEvent getNextNonBlocked() throws Exception;
+	@Deprecated
+	default BufferOrEvent getNextNonBlocked() throws Exception {
+		Optional<BufferOrEvent> bufferOrEvent = pollNext();
```

Comment:
I am a bit confused: this does not look like a blocking implementation (e.g. using `isAvailable()`) but rather like it could actually violate the conditions on the next line.